### PR TITLE
NullDeref check changed from injection of if(stmt) to a runtime call.

### PR DESCRIPTION
**Pull request by CristinaCristescu**

### CODE REVIEW

File: interpreter/cling/include/cling/Interpreter/RuntimeUniverse.h [hunk 1]
```diff
@@ -175,6 +175,11 @@ extern "C" {
   /// definition of the exceptions from the RuntimeUniverse and allows us to
   /// run cling in -no-rtti mode.
   ///
+
+  void cling__runtime__internal__check__and__throw__if__invalid(void* Sema,
```
Comment (karies) [inline interpreter/cling/include/cling/Interpreter/RuntimeUniverse.h]: The `__` was meant to signal `::` in C :-) Maybe rename it to `cling_runtime_internal_ThrowOnInvalidPointer`?

File: interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp [hunk 2]
```diff
@@ -213,18 +213,18 @@ namespace cling {
...
 
       Scope* S = m_Sema.getScopeForContext(m_Sema.CurContext);
       DeclarationName Name
-        = &Context.Idents.get("cling__runtime__internal__throwNullDerefException");
+        = &Context.Idents.get("cling__runtime__internal__check__and__throw__if__invalid");
 
       SourceLocation noLoc;
       LookupResult R(m_Sema, Name, noLoc, Sema::LookupOrdinaryName,
                    Sema::ForRedeclaration);
       m_Sema.LookupQualifiedName(R, Context.getTranslationUnitDecl());
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp]: That lookup shouldn't change - once we found it we found it. Can you store it as a member, such that we only need to do the lookup once?

File: interpreter/cling/lib/Interpreter/Exception.cpp [hunk 3]
```diff
@@ -13,8 +13,18 @@
...
 extern "C" {
+void cling__runtime__internal__check__and__throw__if__invalid(void* Sema, void* Expr, const void* Arg) {
+  if (Arg)
+    return;
+  clang::Sema* S = (clang::Sema*)Sema;
+  clang::Expr* E = (clang::Expr*)Expr;
+  throw cling::NullDerefException(S, E);
+}
+
 void cling__runtime__internal__throwNullDerefException(void* Sema, void* Expr) {
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/Exception.cpp]: Do we still need cling__runtime__internal__throwNullDerefException()?

File: interpreter/cling/include/cling/Interpreter/RuntimeUniverse.h (hunk 1 above)
Comment (CristinaCristescu) [inline interpreter/cling/include/cling/Interpreter/RuntimeUniverse.h]: Done!

File: interpreter/cling/lib/Interpreter/Exception.cpp (hunk 3 above)
Comment (CristinaCristescu) [inline interpreter/cling/lib/Interpreter/Exception.cpp]: Not really, I grep-ed for it and removed it from RequiredSymbols, that the only place I found it.

**Comment by karies** [comment]
Thanks, merged!

**Pull request closed without merging**